src/main/java/duke/Duke.java

```diff
@@ -0,0 +1,184 @@
+                    break;
+                case FIND:
+                    this.ui.showSearchResult(this.taskList, Parser.parseArgs(userInput));
+                    break;
+                }
+            } catch (DukeException e) {
+                this.ui.showDukeException(e);
+            }
+        }
+        this.ui.showFarewell();
+    }
+
+
+    private void exit() {
+        try {
+            this.ui.closeInput();
+            this.storage.saveData(this.taskList);
+        } catch (IOException e) {
+            this.ui.showSavingError();
+        }
+    }
+
+
+    private void addTodo(String args) throws DukeMissingArgumentException {
+        try {
+            Task todo = new Todo(args);
+            this.taskList.add(todo);
+            this.ui.showAdd(todo, this.taskList.getLength());
+        } catch (IndexOutOfBoundsException e) {
+            throw new DukeMissingArgumentException();
+        }
+    }
+
+
+    private void addEvent(String args) throws DukeMissingArgumentException {
+        try {
+            DateTimeFormatter formatter = DateTimeFormatter.ofPattern("dd/MM/yy HH:mm");
+            String[] splits = args.split(" /from ", 2);
+            String[] timestamps = splits[1].split(" /to ", 2);
+            String start = timestamps[0];
+            String end = timestamps[1];
+            LocalDateTime startTime = LocalDateTime.parse(start, formatter);
+            LocalDateTime endTime = LocalDateTime.parse(end, formatter);
+            if (startTime.isAfter(endTime)) {
+                this.ui.showInvalidDateRange();
+                return;
+            }
+
+            Task event = new Event(splits[0], startTime, endTime);
+            this.taskList.add(event);
+            this.ui.showAdd(event, this.taskList.getLength());
+        } catch (IndexOutOfBoundsException e) {
+            throw new DukeMissingArgumentException();
+        } catch (DateTimeParseException e) {
+            this.ui.showInvalidDateFormat();
+        }
+    }
+
+
+    private void addDeadline(String args) throws DukeMissingArgumentException {
+        try {
+            DateTimeFormatter formatter = DateTimeFormatter.ofPattern("dd/MM/yy HH:mm");
+            String[] splits = args.split(" /by ", 2);
+            LocalDateTime time = LocalDateTime.parse(splits[1], formatter);
+
+            Task deadline = new Deadline(splits[0], time);
+            this.taskList.add(deadline);
+            this.ui.showAdd(deadline, this.taskList.getLength());
+        } catch (IndexOutOfBoundsException e) {
+            throw new DukeMissingArgumentException();
+        } catch (DateTimeParseException e) {
+            this.ui.showInvalidDateFormat();
+        }
+    }
+
+
+    private void markAsDone(String args)
+            throws DukeNoTaskFoundException, DukeMissingArgumentException, DukeInvalidArgumentException {
+        try {
+            int taskNum = Integer.parseInt(args);
+            if (taskNum > this.taskList.getLength()) {
```

Comment:
Should we check if taskNum is a negative integer as well?

Reply:
Yep I completely forgot about this! Thanks for the catch!